`import.meta.filename` and `import.meta.dirname` is `undefined` on Node.js Latest LTS `v20.11.0`

### Describe the bug

Node.js implemented API that `import.meta.filename` and `import.meta.dirname` on `v21.2.0`,
and it backported into latest LTS `v20.11.0`.

- https://github.com/nodejs/node/pull/48740

- https://nodejs.org/en/blog/release/v21.2.0
- https://nodejs.org/en/blog/release/v20.11.0



### Reproduction

vitest: `v1.2.1`

Repo: https://github.com/d-hajime-san/reference-vitest-import-meta-object

Result

```bash
% pnpm test

> reference-vitest-import-meta-object@ test /Users/hostname/repository/reference-vitest-import-meta-object
> vitest run


 RUN  v1.2.1 /Users/hostname/repository/reference-vitest-import-meta-object

 ❯ src/module.test.ts (2)
   ❯ testing import.meta (2)
     × filename
     × dirname

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯ Failed Tests 2 ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯

 FAIL  src/module.test.ts > testing import.meta > filename
AssertionError: expected undefined to be '/Users/hostname/repository/ref…' // Object.is equality

- Expected: 
"/Users/hostname/repository/reference-vitest-import-meta-object/src/module.test.ts"

+ Received: 
undefined

 ❯ src/module.test.ts:10:34
      8| describe("testing import.meta", () => {
      9|   it("filename", () => {
     10|     expect(import.meta.filename).toBe(__filename);
       |                                  ^
     11|   });
     12|   it("dirname", () => {

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/2]⎯

 FAIL  src/module.test.ts > testing import.meta > dirname
AssertionError: expected undefined to be '/Users/hostname/repository/ref…' // Object.is equality

- Expected: 
"/Users/hostname/repository/reference-vitest-import-meta-object/src"

+ Received: 
undefined

 ❯ src/module.test.ts:13:33
     11|   });
     12|   it("dirname", () => {
     13|     expect(import.meta.dirname).toBe(__dirname);
       |                                 ^
     14|   });
     15| });

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[2/2]⎯

 Test Files  1 failed (1)
      Tests  2 failed (2)
   Start at  17:09:53
   Duration  247ms (transform 49ms, setup 0ms, collect 39ms, tests 5ms, environment 0ms, prepare 85ms)

 ELIFECYCLE  Test failed. See above for more details
```

### System Info

```shell
System:
  OS: macOS 13.5.1
  CPU: (8) arm64 Apple M1
  Memory: 588.38 MB / 16.00 GB
  Shell: 5.9 - /bin/zsh
Binaries:
  Node: 20.11.0 - ~/.volta/tools/image/node/20.11.0/bin/node
  npm: 10.2.4 - ~/.volta/tools/image/node/20.11.0/bin/npm
  pnpm: 8.14.0 - ~/.volta/bin/pnpm
Browsers:
  Chrome: 120.0.6099.234
  Chrome Canary: 122.0.6256.6
  Edge: 120.0.2210.144
  Safari: 16.6
  Safari Technology Preview: 17.0
npmPackages:
  vitest: 1.2.1 => 1.2.1
```


### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/package.json
+++ b/package.json
@@ -43,7 +43,7 @@
     "@rollup/plugin-commonjs": "^25.0.7",
     "@rollup/plugin-json": "^6.0.1",
     "@rollup/plugin-node-resolve": "^15.2.3",
-    "@types/node": "^18.18.9",
+    "@types/node": "^20.11.5",
     "@types/ws": "^8.5.9",
     "@vitest/browser": "workspace:*",
     "@vitest/coverage-istanbul": "workspace:*",

--- a/packages/vite-node/src/client.ts
+++ b/packages/vite-node/src/client.ts
@@ -296,7 +296,14 @@ export class ViteNodeRunner {
     const modulePath = cleanUrl(moduleId)
     // disambiguate the `<UNIT>:/` on windows: see nodejs/node#31710
     const href = pathToFileURL(modulePath).href
-    const meta = { url: href, env }
+    const __filename = fileURLToPath(href)
+    const __dirname = dirname(__filename)
+    const meta = {
+      url: href,
+      env,
+      filename: __filename,
+      dirname: __dirname,
+    }
     const exports = Object.create(null)
     Object.defineProperty(exports, Symbol.toStringTag, {
       value: 'Module',
@@ -344,7 +351,6 @@ export class ViteNodeRunner {
     })
 
     Object.assign(mod, { code: transformed, exports })
-    const __filename = fileURLToPath(href)
     const moduleProxy = {
       set exports(value) {
         exportAll(cjsExports, value)
@@ -388,7 +394,7 @@ export class ViteNodeRunner {
       exports: cjsExports,
       module: moduleProxy,
       __filename,
-      __dirname: dirname(__filename),
+      __dirname,
     })
 
     debugExecute(__filename)

--- a/packages/vitest/package.json
+++ b/packages/vitest/package.json
@@ -173,6 +173,7 @@
     "@types/istanbul-reports": "^3.0.4",
     "@types/jsdom": "^21.1.6",
     "@types/micromatch": "^4.0.6",
+    "@types/node": "^20.11.5",
     "@types/prompts": "^2.4.9",
     "@types/sinonjs__fake-timers": "^8.1.5",
     "birpc": "0.2.14",

--- a/packages/vitest/src/runtime/external-executor.ts
+++ b/packages/vitest/src/runtime/external-executor.ts
@@ -71,11 +71,11 @@ export class ExternalModulesExecutor {
   }
 
   public resolveModule = async (specifier: string, referencer: string) => {
-    const identifier = await this.resolve(specifier, referencer)
+    const identifier = this.resolve(specifier, referencer)
     return await this.createModule(identifier)
   }
 
-  public async resolve(specifier: string, parent: string) {
+  public resolve(specifier: string, parent: string) {
     for (const resolver of this.resolvers) {
       const id = resolver(specifier, parent)
       if (id)

--- a/packages/vitest/src/runtime/vm/commonjs-executor.ts
+++ b/packages/vitest/src/runtime/vm/commonjs-executor.ts
@@ -76,7 +76,7 @@ export class CommonjsExecutor {
         return _require
       }
 
-      static register() {
+      static register = () => {
         throw new Error(`[vitest] "register" is not available when running in Vitest.`)
       }
 

--- a/packages/vitest/src/runtime/vm/commonjs-executor.ts
+++ b/packages/vitest/src/runtime/vm/commonjs-executor.ts
@@ -76,7 +76,7 @@ export class CommonjsExecutor {
         return _require
       }
 
-      register() {
+      static register() {
         throw new Error(`[vitest] "register" is not available when running in Vitest.`)
       }
 

--- a/packages/vitest/src/runtime/vm/commonjs-executor.ts
+++ b/packages/vitest/src/runtime/vm/commonjs-executor.ts
@@ -76,6 +76,10 @@ export class CommonjsExecutor {
         return _require
       }
 
+      register() {
+        throw new Error(`[vitest] "register" is not available when running in Vitest.`)
+      }
+
       _compile(code: string, filename: string) {
         const cjsModule = Module.wrap(code)
         const script = new vm.Script(cjsModule, {

--- a/packages/vitest/src/runtime/vm/esm-executor.ts
+++ b/packages/vitest/src/runtime/vm/esm-executor.ts
@@ -66,8 +66,8 @@ export class EsmExecutor {
           meta.url = mod.identifier
           if (mod.identifier.startsWith('file:')) {
             const filename = fileURLToPath(mod.identifier)
-            mod.filename = filename
-            mod.dirname = dirname(filename)
+            meta.filename = filename
+            meta.dirname = dirname(filename)
           }
           meta.resolve = (specifier: string, importer?: string | URL) => {
             return this.executor.resolve(specifier, importer != null ? importer.toString() : mod.identifier)

--- a/packages/vitest/src/runtime/vm/esm-executor.ts
+++ b/packages/vitest/src/runtime/vm/esm-executor.ts
@@ -1,6 +1,8 @@
 /* eslint-disable antfu/no-cjs-exports */
 
 import type vm from 'node:vm'
+import { fileURLToPath } from 'node:url'
+import { dirname } from 'node:path'
 import type { ExternalModulesExecutor } from '../external-executor'
 import type { VMModule } from './types'
 import { SourceTextModule, SyntheticModule } from './utils'
@@ -62,8 +64,13 @@ export class EsmExecutor {
         importModuleDynamically: this.executor.importModuleDynamically,
         initializeImportMeta: (meta, mod) => {
           meta.url = mod.identifier
-          meta.resolve = (specifier: string, importer?: string) => {
-            return this.executor.resolve(specifier, importer ?? mod.identifier)
+          if (mod.identifier.startsWith('file:')) {
+            const filename = fileURLToPath(mod.identifier)
+            mod.filename = filename
+            mod.dirname = dirname(filename)
+          }
+          meta.resolve = (specifier: string, importer?: string | URL) => {
+            return this.executor.resolve(specifier, importer != null ? importer.toString() : mod.identifier)
           }
         },
       },

--- a/test/core/test/file-path.test.ts
+++ b/test/core/test/file-path.test.ts
@@ -5,6 +5,14 @@ import { isWindows, slash, toFilePath } from '../../../packages/vite-node/src/ut
 vi.mock('fs')
 
 describe('current url', () => {
+  it('__filename is equal to import.meta.url', () => {
+    expect(__filename).toEqual(import.meta.filename)
+  })
+
+  it('__dirname is equal to import.meta.dirname', () => {
+    expect(__dirname).toEqual(import.meta.dirname)
+  })
+
   describe.runIf(!isWindows)('unix', () => {
     it('__filename', () => {
       expect(__filename.startsWith('file://')).toBe(false)

